File: systems-experiments/flink-experiments/src/main/java/flink_experiments/operators/BinarySource.java
package flink_experiments.operators;

import java.util.concurrent.TimeUnit;

import org.apache.flink.streaming.api.functions.source.SourceFunction;

import flink_experiments.messages.BinaryData;

public class BinarySource implements SourceFunction<BinaryData> {
    private boolean is_running = true;

    private final int msg_size;
    private final int num_warmup_msgs;
    private final int num_msgs;
    private final double send_frequency;

    public BinarySource(final int msg_size, final int num_warmup_msgs, final int num_msgs,
            final double send_frequency) {
        this.msg_size = msg_size;
        this.num_warmup_msgs = num_warmup_msgs;
        this.num_msgs = num_msgs;
        this.send_frequency = send_frequency;
    }

    private void send(long timestamp, SourceContext<BinaryData> ctx) {
        BinaryData data = new BinaryData(this.msg_size);
        data.update_send_time();
        ctx.collectWithTimestamp(data, timestamp);
    }

    private void send_messages(int num_msgs, long start_timestamp, SourceContext<BinaryData> ctx) throws Exception {
        double period = 1.0 / send_frequency;

        long send_time = System.nanoTime();
        for (long timestamp = start_timestamp; this.is_running && timestamp < start_timestamp + num_msgs; timestamp++) {
            this.send(timestamp, ctx);

            long elapsed_time = System.nanoTime() - send_time;
            if (elapsed_time < (long) (period * 1e9)) {
                // Sleep until the next send time.
                TimeUnit.NANOSECONDS.sleep((long) (period * 1e9) - elapsed_time);
            } else {
                System.out.print("BinarySource: fell behind by ");
                System.out.print(Long.toString(elapsed_time - (long) (period * 1e9)));
                System.out.println(" ns");
            }

            send_time += (long) (period * 1e9);
        }

    }

    @Override
    public void run(SourceContext<BinaryData> ctx) throws Exception {

        System.out.println("Starting warmup");
        this.send_messages(this.num_warmup_msgs, 0, ctx);
        System.out.println("Ended warmup");

        System.out.println("Sending messages");
        this.send_messages(this.num_msgs, (long) this.num_warmup_msgs, ctx);
        System.out.println("Done sending messages");
    }

    @Override
    public void cancel() {
        this.is_running = false;
    }
}

File: systems-experiments/erdos-experiments/src/operators/binary_source.rs
use std::{
    thread,
    time::{Duration, Instant},
};

use erdos::{
    dataflow::{
        message::*, operator::Source, stream::errors::SendError, stream::WriteStreamT, Timestamp,
        WriteStream,
    },
    OperatorConfig,
};

use crate::messages::BinaryData;

pub struct BinarySource {
    msg_size: usize,
    num_warmup_msgs: usize,
    num_msgs: usize,
    send_frequency: f32,
}

impl BinarySource {
    pub fn new(
        msg_size: usize,
        num_warmup_msgs: usize,
        num_msgs: usize,
        send_frequency: f32,
    ) -> Self {
        Self {
            msg_size,
            num_warmup_msgs,
            num_msgs,
            send_frequency,
        }
    }

    fn send(
        &self,
        t: Vec<u64>,
        write_stream: &mut WriteStream<BinaryData>,
    ) -> Result<(), SendError> {
        let timestamp = Timestamp::Time(t);
        let data = BinaryData::new(self.msg_size);
        let msg = Message::new_message(timestamp.clone(), data);
        write_stream.send(msg).unwrap();
        write_stream.send(Message::new_watermark(timestamp))
    }
}

impl Source<(), BinaryData> for BinarySource {
    fn run(&mut self, config: &OperatorConfig, write_stream: &mut WriteStream<BinaryData>) {
        let period = Duration::from_secs_f32(1.0 / self.send_frequency);

        tracing::info!("{}: starting warmup", config.get_name());

        let mut send_time = Instant::now();
        for t in 0..self.num_warmup_msgs {
            self.send(vec![0, t as u64], write_stream).unwrap();
            if send_time.elapsed() < period {
                // Sleep until the next send time.
                // This should be fine for high frequencies if the OS supports nanosleep.
                thread::sleep(period - send_time.elapsed());
            } else {
                tracing::warn!(
                    "{}: fell behind by {:?}",
                    config.get_name(),
                    send_time.elapsed() - period
                )
            }
            send_time += period;
        }
        tracing::info!("{}: ended warmup", config.get_name());

        tracing::info!("{}: sending messages", config.get_name());
        for t in 1..self.num_msgs + 1 {
            self.send(vec![t as u64, self.num_warmup_msgs as u64], write_stream)
                .unwrap();
            if send_time.elapsed() < period {
                // Sleep until the next send time.
                // This should be fine for high frequencies if the OS supports nanosleep.
                thread::sleep(period - send_time.elapsed());
            } else {
                tracing::warn!(
                    "{}: fell behind by {:?}",
                    config.get_name(),
                    send_time.elapsed() - period
                )
            }
            send_time += period;
        }

        tracing::debug!("Done sending messages.");

        // Send top watermark to indicate throughput measurements are done.
        write_stream
            .send(Message::new_watermark(Timestamp::Top))
            .unwrap();
    }
}
